### PR TITLE
test(add-new-dc): Add new dc test

### PR DESCRIPTION
**Pull request by soyacz**
trello: https://trello.com/c/v0fwzr67 and https://trello.com/c/DNqa9ddm
jankins job (staging): https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/add-new-dc/
## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [x] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [x] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by soyacz** [comment]
I've tried to follow as closely as possible, but some steps might not be visible in the test flow or are redundant. Here are the details of skipped steps:
1. First step from procedure is actually not working: https://github.com/scylladb/scylla-docs/pull/3837
2. no change in snitch required as we already use Ec2MultiRegionSnitch
3. Also no need to change DC and Rack names - things are already properly done by sct
4. Due to that - no restart is required in existing datacenter
5. this step is properly done when using add_node() method in sct
6. Same as above
7. step done by sct by new_node()
8. done in test flow
9. done in test flow
10. done in test flow
11. done in test flow
12. skipped - new nodes are not seed nodes, I can adapt the code to include that (this will require changing existing nodes configs and restarting all the nodes)
13. There's a function `self.monitors.reconfigure_scylla_monitoring()` - I've used it but not verifying correctness in the test.

Also stress commands are pretty easy on cluster - I keep them small for development.

@roydahan let me know what do you think

File: test-cases/features/add-new-dc.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+n_db_nodes: 3 0  # make n_db_nodes configured as multi-dc with last dc set to 0 (so later easily new node can be added)
```
Comment (fgelcer) [inline test-cases/features/add-new-dc.yaml:1]: you could start this test with a multiDC cluster (like `n_db_nodes: 3 3`) and then add an extra DC

File: test-cases/features/add-new-dc.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+region_name: 'eu-west-1 eu-west-2'
+n_loaders: 1
+n_monitor_nodes: 1
+
+instance_type_db: 'i3.xlarge'
+seeds_num: 3
```
Comment (fgelcer) [inline test-cases/features/add-new-dc.yaml:7]: why do you have 3 seed nodes?

File: test-cases/features/add-new-dc.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
```
Comment (fgelcer) [inline test-cases/features/add-new-dc.yaml:8]: where is the load definition?

File: add_new_dc_test.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+
+class TestAddNewDc(LongevityTest):
+    """Test for procedure:
+     https://docs.scylladb.com/operating-scylla/procedures/cluster-management/add-dc-to-existing-dc/
+     """
+
+    def test_add_new_dc(self) -> None:  # pylint: disable=too-many-locals
+
+        assert self.params.get('n_db_nodes').endswith(" 0"), "n_db_nodes must be a list and last dc must equal 0"
```
Comment (fgelcer) [inline add_new_dc_test.py:22]: why? you could have 2 DCs, and this test will add a 3rd one

File: add_new_dc_test.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        system_keyspaces = ["system_auth", "system_distributed", "system_traces"]
+        self.reconfigure_keyspaces_to_use_network_topology_strategy(
+            keyspaces=system_keyspaces,
+            replication_factors={"eu-westscylla_node_west": 3}
+        )
+        self.prewrite_db_with_data()
```
Comment (fgelcer) [inline add_new_dc_test.py:28]: this function could be taking the c-s load from the test config file

File: add_new_dc_test.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        read_thread, write_thread = self.start_stress_during_adding_new_dc()
```
Comment (fgelcer) [inline add_new_dc_test.py:29]: same here

File: add_new_dc_test.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+                           f"-node {node}"
+        read_thread = self.run_stress_thread(stress_cmd=stress_cmd_read, stats_aggregate_cmds=False, round_robin=False)
+        write_thread = self.run_stress_thread(stress_cmd=stress_cmd_write,
+                                              stats_aggregate_cmds=False, round_robin=False)
+        self.log.info("Stress during adding DC started")
+        return read_thread, write_thread
+
+    def add_node_in_new_dc(self) -> BaseNode:
+        self.log.info("Adding new node")
+        new_node = self.db_cluster.add_nodes(1, dc_idx=1, enable_auto_bootstrap=True)[0]  # add node
```
Comment (fgelcer) [inline add_new_dc_test.py:91]: in this case, you are adding a new node, to an existing (but empty) DC...
the idea is to create a new DC, and add a new node to it

File: add_new_dc_test.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+     https://docs.scylladb.com/operating-scylla/procedures/cluster-management/add-dc-to-existing-dc/
+     """
+
+    def test_add_new_dc(self) -> None:  # pylint: disable=too-many-locals
+
+        assert self.params.get('n_db_nodes').endswith(" 0"), "n_db_nodes must be a list and last dc must equal 0"
+        system_keyspaces = ["system_auth", "system_distributed", "system_traces"]
+        self.reconfigure_keyspaces_to_use_network_topology_strategy(
+            keyspaces=system_keyspaces,
+            replication_factors={"eu-westscylla_node_west": 3}
```
Comment (fgelcer) [inline add_new_dc_test.py:26]: the DC name could be found dynamically

File: add_new_dc_test.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        self.prewrite_db_with_data()
+        read_thread, write_thread = self.start_stress_during_adding_new_dc()
+        # no need to change network topology
+        new_node = self.add_node_in_new_dc()
+
+        self.querying_new_node_should_return_no_data(new_node)  # verify issue #8354
+
+        self.reconfigure_keyspaces_to_use_network_topology_strategy(
+            keyspaces=system_keyspaces + ["keyspace1"],
+            replication_factors={"eu-westscylla_node_west": 3, "eu-west-2scylla_node_west": 1}
```
Comment (fgelcer) [inline add_new_dc_test.py:37]: the DC names could be found dynamically

File: add_new_dc_test.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        )
+
+        self.log.info("Running rebuild on each node in new DC")
+        new_node.run_nodetool(sub_cmd="rebuild -- eu-westscylla_node_west")
```
Comment (fgelcer) [inline add_new_dc_test.py:41]: same here

Comment (fgelcer) [reply add_new_dc_test.py:41]: not better to add `publish_event=True` to this command?

File: add_new_dc_test.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+    def reconfigure_keyspaces_to_use_network_topology_strategy(self, keyspaces: List[str], replication_factors: dict[str, int]) -> None:
+        node = self.db_cluster.nodes[0]
+        self.log.info("Reconfiguring keyspace Replication Strategy")
+        network_topology_strategy = NetworkTopologyReplicationStrategy(
+            **replication_factors)
+        for keyspace in keyspaces:
+            cql = f"ALTER KEYSPACE {keyspace} WITH replication = {network_topology_strategy}"
+            node.run_cqlsh(cql)
+        self.log.info("Replication Strategies for {} reconfigured".format(keyspaces))
```
Comment (fgelcer) [inline add_new_dc_test.py:62]: recommended to use lazy interpolation:
`self.log.info("Replication Strategies for %s reconfigured", keyspaces)`

File: add_new_dc_test.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+    def prewrite_db_with_data(self) -> None:
```
Comment (fgelcer) [inline add_new_dc_test.py:64]: should be defined in the test config file

File: add_new_dc_test.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        self.log.info("Prewriting database...")
+        stress_cmd = "cassandra-stress write cl=QUORUM n=209 -schema " \
+                     "'replication(strategy=NetworkTopologyStrategy,eu-westscylla_node_west=3) " \
+                     "compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native " \
+                     "-rate threads=8 -pop seq=1..209 -col 'n=FIXED(10) size=FIXED(512)' -log interval=5"
+        pre_thread = self.run_stress_thread(stress_cmd=stress_cmd, stats_aggregate_cmds=False, round_robin=False)
+        self.verify_stress_thread(cs_thread_pool=pre_thread)
+        self.log.info("Database pre write completed")
+
+    def start_stress_during_adding_new_dc(self) -> Tuple[CassandraStressThread, CassandraStressThread]:
```
Comment (fgelcer) [inline add_new_dc_test.py:74]: should be defined in the test config file

File: add_new_dc_test.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+        self.db_cluster.wait_for_nodes_up_and_normal(nodes=[new_node])
+        self.monitors.reconfigure_scylla_monitoring()
+
+        status = self.db_cluster.get_nodetool_status()
+        assert [dc for dc in list(status.keys()) if dc ==
+                "eu-west-2scylla_node_west"], f"new datacenter was not registered. Cluster status: {status}"
+        self.log.info("New DC to cluster has been added")
+        return new_node
+
+    def verify_data_can_be_read_from_new_dc(self, new_node: BaseNode) -> None:
```
Comment (fgelcer) [inline add_new_dc_test.py:103]: should be defined in the test config file

File: add_new_dc_test.py (hunk 4 above)
Comment (soyacz) [reply add_new_dc_test.py:22]: adding another dc with 0 makes everything much easier. Without it I'd need to mess with undefined number of things to setup by myself in the code, just to make sct happy (e.g. _ec2_subnet_id, _ec2_security_group_ids, _ec2_ami_id, credentials and possibly many more).

File: add_new_dc_test.py (hunk 7 above)
Comment (soyacz) [reply add_new_dc_test.py:91]: from cluster perspective, this DC does not exist. Adding DC with 0 nodes shouldn't affect existing DC's nodes in any way. This approach makes everything much easier and more future proof (when sct changes).

File: test-cases/features/add-new-dc.yaml (hunk 2 above)
Comment (soyacz) [reply test-cases/features/add-new-dc.yaml:7]: I didn't pay attention to that. How much it should be?

File: test-cases/features/add-new-dc.yaml (hunk 1 above)
Comment (soyacz) [reply test-cases/features/add-new-dc.yaml]: that was intended - for dev purposes it was only 1 starting DC. Fixed.

File: test-cases/features/add-new-dc.yaml (hunk 3 above)
Comment (soyacz) [reply test-cases/features/add-new-dc.yaml:15]: added to yaml.

File: add_new_dc_test.py (hunk 11 above)
Comment (soyacz) [reply add_new_dc_test.py:66]: removed, possibly it is not needed (left after investigation).

**Comment by soyacz** [comment]
Shall we backport it to 4.6? any other version?

File: add_new_dc_test.py (hunk 4 above)
Comment (roydahan) [reply add_new_dc_test.py:22]: The more interesting scenario is growing from 1 DC to 2 DCs.

**Comment by roydahan** [comment]
> I've tried to follow as closely as possible, but some steps might not be visible in the test flow or are redundant. Here are the details of skipped steps:
> 
> 1. First step from procedure is actually not working: [fix add new dc procedure scylla-docs#3837](https://github.com/scylladb/scylla-docs/pull/3837)
> 2. no change in snitch required as we already use Ec2MultiRegionSnitch
> 3. Also no need to change DC and Rack names - things are already properly done by sct
> 4. Due to that - no restart is required in existing datacenter
> 5. this step is properly done when using add_node() method in sct
> 6. Same as above
> 7. step done by sct by new_node()
> 8. done in test flow
> 9. done in test flow
> 10. done in test flow
> 11. done in test flow
> 12. skipped - new nodes are not seed nodes, I can adapt the code to include that (this will require changing existing nodes configs and restarting all the nodes)
> 13. There's a function `self.monitors.reconfigure_scylla_monitoring()` - I've used it but not verifying correctness in the test.
> 
> For now test starts with one DC - for faster development - can be changed quite easily. And stress commands are pretty easy on cluster now - I keep them small for development.
> 
> @roydahan let me know what do you think

Overall looks good to me.
I'm thinking maybe we should not rely in this test on the default provisioning that SCT does and somehow change it, so the test will test also steps 2-4.

**Comment by soyacz** [comment]
> I'm thinking maybe we should not rely in this test on the default provisioning that SCT does and somehow change it, so the test will test also steps 2-4.

Replacing SCT provisioning with own procedure won't be easy and fast.
@roydahan is it worth the effort?

File: add_new_dc_test.py (hunk 4 above)
Comment (soyacz) [reply add_new_dc_test.py:22]: I was trying to follow Trello task - where is written we should start with 2. 
I'd be happy to change it back to scenario 1DC->2DC.

**Comment by roydahan** [comment]
> > I'm thinking maybe we should not rely in this test on the default provisioning that SCT does and somehow change it, so the test will test also steps 2-4.
> 
> Replacing SCT provisioning with own procedure won't be easy and fast. @roydahan is it worth the effort?

Probably not, at least not for now.

**Review by roydahan — APPROVED** [review]

**Comment by fgelcer** [comment]
@roydahan , 1st there were few unaddressed comments...
other thing is, what about back ports? i think it should be back ported to 4.6

**Comment by roydahan** [comment]
The comments weren’t crucial.
The backport happens automatically, but creating the job and putting it in the spreadsheet doesn’t.
I suggest to run it on master, and if works as we want, copy it to 4.6.

**Pull request merged**